### PR TITLE
Lazy allocate dictionary in ItemsDictionary

### PR DESCRIPTION
**Pull request by benaadams**
Lazy allocate dictionary in ItemsDictionary (doesn't need to allocate a dict to find out an item is not there)

### CODE REVIEW

File: src/Http/Http/src/Internal/ItemsDictionary.cs [hunk 1]
```diff
@@ -8,111 +8,160 @@ namespace Microsoft.AspNetCore.Http.Internal
...
+            value = null;
+            return _items != null && _items.TryGetValue(key, out value);
         }
 
         ICollection<object> IDictionary<object, object>.Values
         {
-            get { return Items.Values; }
+            get
+            {
+                if (_items == null)
```
Comment (Tratcher) [inline src/Http/Http/src/Internal/ItemsDictionary.cs]:
```
EnsureDictionary()
{
 if (_items == null)...
}
```

Comment (benaadams) [reply src/Http/Http/src/Internal/ItemsDictionary.cs]: Done

File: src/Http/Http/src/Internal/ItemsDictionary.cs [hunk 2]
```diff
@@ -8,111 +8,147 @@ namespace Microsoft.AspNetCore.Http.Internal
...
+            => _items != null && _items.Contains(item);
 
         void ICollection<KeyValuePair<object, object>>.CopyTo(KeyValuePair<object, object>[] array, int arrayIndex)
         {
-            Items.CopyTo(array, arrayIndex);
-        }
+            if (_items == null)
+            {
+                //Delegate to Empty Dictionary to do the argument checking.
+                EmptyDictionary.Collection.CopyTo(array, arrayIndex);
```
Comment (Tratcher) [inline src/Http/Http/src/Internal/ItemsDictionary.cs]: If the collection is empty then isn't this a no-op?

File: src/Http/Http/src/Internal/ItemsDictionary.cs [hunk 3]
```diff
@@ -8,111 +8,147 @@ namespace Microsoft.AspNetCore.Http.Internal
...
+            value = null;
+            return _items != null && _items.TryGetValue(key, out value);
         }
 
         ICollection<object> IDictionary<object, object>.Values
         {
-            get { return Items.Values; }
+            get
+            {
+                EnsureDictionary();
```
Comment (davidfowl) [inline src/Http/Http/src/Internal/ItemsDictionary.cs]: Why allocate here if the inner dictionary is null?

File: src/Http/Http/src/Internal/ItemsDictionary.cs [hunk 4]
```diff
@@ -8,111 +8,147 @@ namespace Microsoft.AspNetCore.Http.Internal
...
-            return Items.ContainsKey(key);
-        }
+            => _items != null && _items.ContainsKey(key);
 
         ICollection<object> IDictionary<object, object>.Keys
         {
-            get { return Items.Keys; }
+            get
+            {
+                EnsureDictionary();
```
Comment (davidfowl) [inline src/Http/Http/src/Internal/ItemsDictionary.cs]: Why allocate here if the inner dictionary is null?

File: src/Http/Http/src/Internal/ItemsDictionary.cs (hunk 3 above)
Comment (benaadams) [reply src/Http/Http/src/Internal/ItemsDictionary.cs]: Mistakenly thought the collections were mutable

Comment (benaadams) [reply src/Http/Http/src/Internal/ItemsDictionary.cs]: Changed

File: src/Http/Http/src/Internal/ItemsDictionary.cs (hunk 4 above)
Comment (benaadams) [reply src/Http/Http/src/Internal/ItemsDictionary.cs]: Changed

**Review by davidfowl — APPROVED** [review]

**Pull request merged**